File: src/Keybindings.cpp
#include "Keybindings.hpp"
#include "Utils.hpp"

// Win32HotkeyRegistrar Implementation

Win32HotkeyRegistrar::Win32HotkeyRegistrar(HWND hwnd) : m_hwnd(hwnd) {}

Win32HotkeyRegistrar::~Win32HotkeyRegistrar() { unregisterAll(); }

void Win32HotkeyRegistrar::registerHotkey(int id, const KeyCombination &combo) {
  if (RegisterHotKey(m_hwnd, id, combo.modifiers, combo.key)) {
    m_registeredIds.push_back(id);
  } else {
    utils::debugLog("Failed to register hotkey with ID {}", id);
  }
}

void Win32HotkeyRegistrar::unregisterAll() {
  for (int id : m_registeredIds) {
    UnregisterHotKey(m_hwnd, id);
  }
  m_registeredIds.clear();
}

// KeybindingManager Implementation

KeybindingManager::KeybindingManager(
    std::unique_ptr<IHotkeyRegistrar> registrar)
    : m_registrar(std::move(registrar)), m_nextId(1) {}

void KeybindingManager::addBinding(const KeyCombination &combo, Action action) {
  int id = m_nextId++;
  m_bindings[id] = action;
  m_registrar->registerHotkey(id, combo);
}

Action KeybindingManager::getActionForId(int id) const {
  auto it = m_bindings.find(id);
  if (it != m_bindings.end()) {
    return it->second;
  }
  return Action::Unknown;
}

void KeybindingManager::initializeDefaults() {
  clear();

  // Default Modifiers
  UINT modDefault = MOD_ALT | MOD_SHIFT | MOD_NOREPEAT;
  UINT modMove = MOD_ALT | MOD_CONTROL | MOD_SHIFT | MOD_NOREPEAT;

  // Arrow keys
  addBinding({modDefault, VK_LEFT}, Action::ScrollLeft);
  addBinding({modDefault, VK_RIGHT}, Action::ScrollRight);
  addBinding({modDefault, VK_UP}, Action::ScrollUp);
  addBinding({modDefault, VK_DOWN}, Action::ScrollDown);
  addBinding({modMove, VK_LEFT}, Action::MoveWindowLeft);
  addBinding({modMove, VK_RIGHT}, Action::MoveWindowRight);
  addBinding({modMove, VK_UP}, Action::MoveWindowUp);
  addBinding({modMove, VK_DOWN}, Action::MoveWindowDown);

  // Vim keys
  addBinding({modDefault, 'H'}, Action::ScrollLeft);
  addBinding({modDefault, 'L'}, Action::ScrollRight);
  addBinding({modDefault, 'K'}, Action::ScrollUp);
  addBinding({modDefault, 'J'}, Action::ScrollDown);
  addBinding({modMove, 'H'}, Action::MoveWindowLeft);
  addBinding({modMove, 'L'}, Action::MoveWindowRight);
  addBinding({modMove, 'K'}, Action::MoveWindowUp);
  addBinding({modMove, 'J'}, Action::MoveWindowDown);

  // Niri columns keys
  addBinding({modMove, VK_OEM_COMMA}, Action::ConsumeOrExpelLeft);
  addBinding({modMove, VK_OEM_PERIOD}, Action::ConsumeOrExpelRight);
  addBinding({modDefault, VK_OEM_COMMA}, Action::ConsumeIntoColumn);
  addBinding({modDefault, VK_OEM_PERIOD}, Action::ExpelFromColumn);

  // Other actions
  addBinding({modDefault, 'E'}, Action::Exit);
  addBinding({modDefault, 'F'}, Action::ToggleFullscreen);
  addBinding({modDefault, 'M'}, Action::MoveToNextMonitor);
}

void KeybindingManager::clear() {
  m_bindings.clear();
  m_registrar->unregisterAll();
  m_nextId = 1;
}

File: src/keybindings.zig
const std = @import("std");
const win32 = @import("win32").everything;

pub const Action = enum {
    ScrollLeft,
    ScrollRight,
    ScrollUp,
    ScrollDown,
    MoveWindowLeft,
    MoveWindowRight,
    MoveWindowUp,
    MoveWindowDown,
    ConsumeOrExpelLeft,
    ConsumeOrExpelRight,
    ConsumeIntoColumn,
    ExpelFromColumn,
    SwitchPresetColumnWidth,
    SwitchPresetWindowHeight,
    FocusWorkspaceDown,
    FocusWorkspaceUp,
    FocusWorkspacePrevious,
    MoveColumnToWorkspaceDown,
    MoveColumnToWorkspaceUp,
    ToggleFullscreen,
    MoveToNextMonitor,
    Exit,
    Unknown,
};

pub const KeyCombination = struct {
    modifiers: win32.HOT_KEY_MODIFIERS,
    key: u32,
};

pub const KeybindingManager = struct {
    bindings: std.AutoHashMap(i32, Action),
    registered_ids: std.ArrayList(i32) = .empty,
    next_id: i32 = 1,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) KeybindingManager {
        return .{
            .bindings = std.AutoHashMap(i32, Action).init(allocator),
            .registered_ids = .empty,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *KeybindingManager) void {
        self.unregisterAll();
        self.bindings.deinit();
        self.registered_ids.deinit(self.allocator);
    }

    pub fn addBinding(self: *KeybindingManager, combo: KeyCombination, action: Action) !void {
        const id = self.next_id;
        self.next_id += 1;
        
        // Passing null for HWND registers global hotkey for this thread
        if (win32.RegisterHotKey(null, id, combo.modifiers, combo.key) != 0) {
            try self.bindings.put(id, action);
            try self.registered_ids.append(self.allocator, id);
        } else {
            std.log.err("Failed to register hotkey with ID {d}. Error: {d}", .{id, win32.GetLastError()});
        }
    }

    pub fn unregisterAll(self: *KeybindingManager) void {
        for (self.registered_ids.items) |id| {
            _ = win32.UnregisterHotKey(null, id);
        }
        self.registered_ids.clearAndFree(self.allocator);
        self.bindings.clearRetainingCapacity();
        self.next_id = 1;
    }

    pub fn getActionForId(self: *const KeybindingManager, id: i32) Action {
        return self.bindings.get(id) orelse Action.Unknown;
    }

    pub fn initializeDefaults(self: *KeybindingManager) !void {
        self.unregisterAll();

        const mod_default = win32.HOT_KEY_MODIFIERS{
            .ALT = 1,
            .SHIFT = 1,
            .NOREPEAT = 1,
        };
        const mod_move = win32.HOT_KEY_MODIFIERS{
            .ALT = 1,
            .CONTROL = 1,
            .SHIFT = 1,
            .NOREPEAT = 1,
        };

        // Arrow keys
        try self.addBinding(.{ .modifiers = mod_default, .key = @intFromEnum(win32.VK_LEFT) }, .ScrollLeft);
        try self.addBinding(.{ .modifiers = mod_default, .key = @intFromEnum(win32.VK_RIGHT) }, .ScrollRight);
        try self.addBinding(.{ .modifiers = mod_default, .key = @intFromEnum(win32.VK_UP) }, .ScrollUp);
        try self.addBinding(.{ .modifiers = mod_default, .key = @intFromEnum(win32.VK_DOWN) }, .ScrollDown);

        try self.addBinding(.{ .modifiers = mod_move, .key = @intFromEnum(win32.VK_LEFT) }, .MoveWindowLeft);
        try self.addBinding(.{ .modifiers = mod_move, .key = @intFromEnum(win32.VK_RIGHT) }, .MoveWindowRight);
        try self.addBinding(.{ .modifiers = mod_move, .key = @intFromEnum(win32.VK_UP) }, .MoveWindowUp);
        try self.addBinding(.{ .modifiers = mod_move, .key = @intFromEnum(win32.VK_DOWN) }, .MoveWindowDown);

        // Vim keys
        try self.addBinding(.{ .modifiers = mod_default, .key = 'H' }, .ScrollLeft);
        try self.addBinding(.{ .modifiers = mod_default, .key = 'L' }, .ScrollRight);
        try self.addBinding(.{ .modifiers = mod_default, .key = 'K' }, .ScrollUp);
        try self.addBinding(.{ .modifiers = mod_default, .key = 'J' }, .ScrollDown);

        try self.addBinding(.{ .modifiers = mod_move, .key = 'H' }, .MoveWindowLeft);
        try self.addBinding(.{ .modifiers = mod_move, .key = 'L' }, .MoveWindowRight);
        try self.addBinding(.{ .modifiers = mod_move, .key = 'K' }, .MoveWindowUp);
        try self.addBinding(.{ .modifiers = mod_move, .key = 'J' }, .MoveWindowDown);

        // Niri columns keys
        try self.addBinding(.{ .modifiers = mod_move, .key = @intFromEnum(win32.VK_OEM_COMMA) }, .ConsumeOrExpelLeft);
        try self.addBinding(.{ .modifiers = mod_move, .key = @intFromEnum(win32.VK_OEM_PERIOD) }, .ConsumeOrExpelRight);
        try self.addBinding(.{ .modifiers = mod_default, .key = @intFromEnum(win32.VK_OEM_COMMA) }, .ConsumeIntoColumn);
        try self.addBinding(.{ .modifiers = mod_default, .key = @intFromEnum(win32.VK_OEM_PERIOD) }, .ExpelFromColumn);

        // Preset bounds keys
        try self.addBinding(.{ .modifiers = mod_default, .key = 'R' }, .SwitchPresetColumnWidth);
        try self.addBinding(.{ .modifiers = mod_move, .key = 'R' }, .SwitchPresetWindowHeight);

        // Workspaces
        try self.addBinding(.{ .modifiers = mod_default, .key = @intFromEnum(win32.VK_NEXT) }, .FocusWorkspaceDown);
        try self.addBinding(.{ .modifiers = mod_default, .key = @intFromEnum(win32.VK_PRIOR) }, .FocusWorkspaceUp);
        try self.addBinding(.{ .modifiers = mod_default, .key = 'U' }, .FocusWorkspaceDown);
        try self.addBinding(.{ .modifiers = mod_default, .key = 'I' }, .FocusWorkspaceUp);
        // try self.addBinding(.{ .modifiers = mod_default, .key = @intFromEnum(win32.VK_TAB) }, .FocusWorkspacePrevious);

        try self.addBinding(.{ .modifiers = mod_move, .key = @intFromEnum(win32.VK_NEXT) }, .MoveColumnToWorkspaceDown);
        try self.addBinding(.{ .modifiers = mod_move, .key = @intFromEnum(win32.VK_PRIOR) }, .MoveColumnToWorkspaceUp);
        try self.addBinding(.{ .modifiers = mod_move, .key = 'U' }, .MoveColumnToWorkspaceDown);
        try self.addBinding(.{ .modifiers = mod_move, .key = 'I' }, .MoveColumnToWorkspaceUp);

        // Other actions
        try self.addBinding(.{ .modifiers = mod_default, .key = 'E' }, .Exit);
        try self.addBinding(.{ .modifiers = mod_default, .key = 'F' }, .ToggleFullscreen);
        try self.addBinding(.{ .modifiers = mod_default, .key = 'M' }, .MoveToNextMonitor);
    }
};
